### PR TITLE
[generate] shape checks in tests compatible with fixed-length caches (+ some minor fixes)

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes a few test cases exposed by #33212 [e.g. some models are failing cache shape checks, if their default cache is not a dynamic cache]. Fix added in a separate PR to avoid making #33212 a huge PR :)

This PR:
1. Updates generate output checks to also work with fixed-length caches;
2. Tests with `output_attentions=True` now explicitly use `eager` attention (`sdpa` doesn't return the attentions, `eager` was being used implicitly with warnings);
3. (enabled by 1. and 2.) Adds a test regarding generate + extra outputs + compile, which was not being tested; 
4. Standardizes the inconsistent method to pull the cache from the model outputs in the generation loop;
5. (enabled by 1.) Deletes unnecessary overwrites.
6. [EDIT, added after PR reviews] updates variable names used in common tests, so we can easily understand what's going on.

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -2083,7 +2083,7 @@ def generate(
         # - `model_kwargs` may be updated in place with a cache as defined by the parameters in `generation_config`.
         # - different models have a different cache name expected by the model (default = "past_key_values")
         # - `max_length`, prepared above, is used to determine the maximum cache length
-        max_cache_length = generation_config.max_length
+        max_cache_length = generation_config.max_length - 1
```
Comment (gante) [inline src/transformers/generation/utils.py:2086]: We were creating caches that were larger than what was set by our max length flags: the last token is not present in the cache.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 2]
```diff
@@ -935,20 +935,21 @@ def forward(
         if self.config.pad_token_id is None and batch_size != 1:
             raise ValueError("Cannot handle batch sizes > 1 if no padding token is defined.")
         if self.config.pad_token_id is None:
-            last_non_pad_token = -1
-        elif input_ids is not None:
-            # To handle both left- and right- padding, we take the rightmost token that is not equal to pad_token_id
-            non_pad_mask = (input_ids != self.config.pad_token_id).to(logits.device, torch.int32)
-            token_indices = torch.arange(input_ids.shape[-1], device=logits.device)
-            last_non_pad_token = (token_indices * non_pad_mask).argmax(-1)
+            sequence_lengths = -1
```
Comment (gante) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:938]: note: the diff in `gptneox`, qwen2.5_vl` and `zamba2` was written by `make fix-copies`!

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -2990,7 +2990,11 @@ def _contrastive_search(
                 next_past_key_values = selected_outputs["past_key_values"]
 
             else:
-                _, next_past_key_values = self._extract_past_from_model_output(outputs)
+                next_past_key_values = None
+                for possible_cache_name in ALL_CACHE_NAMES:
+                    if possible_cache_name in outputs:
+                        next_past_key_values = getattr(outputs, possible_cache_name)
+                        break
```
Comment (ydshieh) [inline src/transformers/generation/utils.py:2997]: maybe can be shorter (nit)

```
next_past_key_values = None
for possible_cache_name in ALL_CACHE_NAMES:
    next_past_key_values = getattr(outputs, possible_cache_name, None)
```

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -2346,21 +2427,22 @@ def _check_logits(self, batch_size, scores, config):
...
+        self.assertEqual(len(attentions), (max_length - min_length))
+
+        has_cache = past_key_values is not None
+        has_static_cache = isinstance(past_key_values, (StaticCache, HybridCache))
 
         for idx, iter_attentions in enumerate(attentions):
-            tgt_len = min_length + idx if not use_cache else 1
-            src_len = min_length + idx
+            tgt_len = min_length + idx if not has_cache else 1
+            src_len = min_length + idx if not has_static_cache else past_key_values.get_max_cache_shape()
```
Comment (ydshieh) [inline tests/generation/test_utils.py:2442]: I had trouble to understand this part (but it's not the new change, I don't understand well what happens here even in the current `main`)

Do you think a bit of comment in the code will help future readers?

File: tests/generation/test_utils.py [hunk 5]
```diff
@@ -56,6 +56,8 @@
     import torch
     import torch.nn.functional as F
 
+    torch._dynamo.config.cache_size_limit = 512  # So we can compile all models with no cache limit exception
```
Comment (ydshieh) [inline tests/generation/test_utils.py:59]: a side question: do we (ever) call `torchdynamo.reset()` or similar method for compilation tests? Would that avoid the `cache limit exception` without set it to `512` here?

File: tests/generation/test_utils.py [hunk 6]
```diff
@@ -2061,6 +2090,56 @@ def test_generate_compile_model_forward(self):
...
+        for model_class in self.all_generative_model_classes:
+            if not model_class._supports_static_cache:
+                self.skipTest("This model doesn't support static cache (= no expectations of compilation support)")
+
+            config, inputs_dict = self.prepare_config_and_inputs_for_generate()
+            if self.has_attentions:
+                config._attn_implementation = "eager"  # can't output attentions otherwise
+            model = model_class(config).to(torch_device).eval()
+
+            has_defined_cache_implementation = model.generation_config.cache_implementation is not None
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:2185]: So far, we don't have any model with a non-default cache that is not compileable, and the model supports compilation. We may need to revisit this condition when that happens, yes :)

(I wouldn't worry about it for now. New caches are usually compileable, otherwise we would need to maintain 2 caches for a specific model if we want that model to be compileable)

File: tests/generation/test_utils.py [hunk 7]
```diff
@@ -2405,45 +2487,41 @@ def _check_encoder_hidden_states_for_generate(self, hidden_states, batch_size, c
             [encoder_expected_shape] * len(hidden_states),
         )
 
-    def _check_past_key_values_for_generate(self, batch_size, past_key_values, seq_length, config, num_beam_groups=1):
-        self.assertIsInstance(past_key_values, (tuple, Cache))
-
-        # Encoder-decoder models: pull and verify the decoder cache
-        if isinstance(past_key_values, EncoderDecoderCache):
-            past_key_values = past_key_values.self_attention_cache
+    def _check_past_key_values_for_generate(self, batch_size, decoder_past_key_values, seq_length, config):
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:2567]: Renaming everywhere 👍 

The name was ambiguous -- these checks need access to the decoder caches, and it is clear after renaming.

File: tests/generation/test_utils.py [hunk 8]
```diff
@@ -2346,21 +2427,22 @@ def _check_logits(self, batch_size, scores, config):
         self.assertTrue(vocab_diff in [0, 1])
         self.assertListEqual([vocab_size - score.shape[-1] for score in scores], [vocab_diff] * len(scores))
 
-    def _check_attentions_for_generate(
-        self, batch_size, attentions, min_length, max_length, config, use_cache=False, num_beam_groups=1
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:2427]: Yes, correct. I have no idea why it was there 👀

File: tests/generation/test_utils.py (hunk 5 above)
Comment (gante) [reply tests/generation/test_utils.py:59]: We can reset the cache in dynamo instead of this, yes. Do you prefer it?

[The advantage of this method would be slightly faster compilation, but the difference should be very small]

Comment (ydshieh) [reply tests/generation/test_utils.py:59]: no, you can keep what it is. I am just thinking in more general: such stuff is always good to be cleaned after each test call (teardown). I have seen similar issue for torchscript and have to add `clear_torch_jit_class_registry` (otherwise memory leak)

File: tests/generation/test_utils.py (hunk 4 above)
Comment (gante) [reply tests/generation/test_utils.py]: These are indeed poor old variable names!

Made the following name updates:
`src_len` -> `query_length`
`tgt_len` -> `model_input_length`
`min_length` -> `prompt_length`
`max_length` -> `output_length`
`idx` -> `generated_length`

I've also resolved other ambiguous names (`seq_len` could mean different things in different places) and slightly rewrote other bits to improve clarity.

I'll ping you again when it's ready for a re-review 🤗

**Comment by gante** [comment]
@ydshieh should be ready for a re-review.

Two notes:
1. `_check_outputs` and especially its inner functions should be much easier to follow. LMK if any part is unclear—documenting these general functions is important to future-proof our test base!
2. As I rewrote the test functions, I noticed some overwrites were not necessary (or at least could be simplified) ✂️

File: tests/generation/test_utils.py [hunk 9]
```diff
@@ -2138,6 +2165,58 @@ def test_generate_compile_model_forward(self):
...
+            if model.config.is_encoder_decoder:
+                self.assertTrue(output_generate.sequences.shape[-1] == self.max_new_tokens + 1)
+                self.assertIsInstance(output_generate, GenerateEncoderDecoderOutput)
+            else:
+                self.assertTrue(
+                    output_generate.sequences.shape[-1] == self.max_new_tokens + inputs_dict["input_ids"].shape[-1]
+                )
+                self.assertIsInstance(output_generate, GenerateDecoderOnlyOutput)
+
+            self._check_outputs(output_generate, model.config, use_cache=True)
```
Comment (ydshieh) [inline tests/generation/test_utils.py:2218]: Maybe it's better to name this like `_check_outputs_for_generate`? WDYT?

File: tests/generation/test_utils.py [hunk 10]
```diff
@@ -2396,131 +2479,141 @@ def _check_outputs(self, output, config, use_cache=False, num_return_sequences=1
...
-        self.assertEqual(len(attentions), (max_length - min_length) * num_beam_groups)
+        self.assertEqual(len(attentions), (output_length - prompt_length))
+
+        use_cache = decoder_past_key_values is not None
+        has_static_cache = isinstance(decoder_past_key_values, (StaticCache, HybridCache))
 
-        for idx, iter_attentions in enumerate(attentions):
-            tgt_len = min_length + idx if not use_cache else 1
-            src_len = min_length + idx
+        for generated_length, iter_attentions in enumerate(attentions):
```
Comment (ydshieh) [inline tests/generation/test_utils.py:2520]: Probably it's obvious to you and @zucchini-nlp , but if you think it's would benefit with a one line comment : 

> `attentions` outputs in a call to `generate` is a list of .... (whatever) .... for each generation step.

**Review by ydshieh — APPROVED** [review]
left 2 nit comments + I don't know well the removal of `_update_model_kwargs_for_generation` in the 2 modeling files (for which I will leave to another reviewer 🙏 )

Otherwise all good to me 💯 thank you

**Review by zucchini-nlp — APPROVED** [review]
Thanks for cleaning up! Left one suggestion for static cache tests :)

File: tests/generation/test_utils.py (hunk 9 above)
Comment (gante) [reply tests/generation/test_utils.py:2218]: 100% agreed, will add 👍

File: tests/generation/test_utils.py (hunk 10 above)
Comment (gante) [reply tests/generation/test_utils.py:2522]: (done)

File: tests/generation/test_utils.py [hunk 11]
```diff
@@ -1765,7 +1794,7 @@ def test_generate_from_inputs_embeds_with_static_cache(self):
             num_hidden_layers = text_config.num_hidden_layers
 
             inputs_embeds = model.get_input_embeddings()(input_ids)
-            max_cache_len += inputs_embeds.shape[1]
+            max_cache_len += inputs_embeds.shape[1] - 1  # the last generated token has no cache
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:1795]: (leaving this one for a separate PR, moving conversation to slack)

File: tests/generation/test_utils.py [hunk 12]
```diff
@@ -2296,80 +2375,84 @@ def _check_outputs(self, output, config, use_cache=False, num_return_sequences=1
...
+        prompt_length = getattr(self.model_tester, "text_seq_length", prompt_length)
 
         config = config.text_config if hasattr(config, "text_config") else config
 
-        gen_len = (
-            output.sequences.shape[-1] - 1 if config.is_encoder_decoder else output.sequences.shape[-1] - seq_length
+        generated_length = (
+            output.sequences.shape[-1] - 1 if config.is_encoder_decoder else output.sequences.shape[-1] - prompt_length
         )
+        decoder_past_key_values = getattr(output, "past_key_values", None)
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:2387]: When `use_cache=False` :P (or models with different cache names, like RWKV)

File: tests/generation/test_utils.py [hunk 13]
```diff
@@ -2396,131 +2479,141 @@ def _check_outputs(self, output, config, use_cache=False, num_return_sequences=1
...
     ):
         self.assertIsInstance(attentions, tuple)
         self.assertListEqual(
             [isinstance(iter_attentions, tuple) for iter_attentions in attentions], [True] * len(attentions)
         )
-        self.assertEqual(len(attentions), (max_length - min_length) * num_beam_groups)
+        self.assertEqual(len(attentions), (output_length - prompt_length))
+
+        use_cache = decoder_past_key_values is not None
+        has_static_cache = isinstance(decoder_past_key_values, (StaticCache, HybridCache))
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:2518]: I'll add a note here, so we can easily know when to do when the test breaks.

(but for now will leave as is, to contain test complexity)
